File: SOLID Principles/Typescript/tests/OCP.test.ts
import { PayProcessor, PaymentProcessor, CreditCardPayment, PayPalPayment, CryptoPayment, ApplePayPayment } from "../OCP";
import type { IPaymentMethod } from "../OCP";
// ─── Bad Implementation Tests ───────────────────────────────────────────────

describe("PayProcessor (Bad Implementation - OCP Violation)", () => {
    let processor: PayProcessor;
    let consoleSpy: jest.SpyInstance;

    beforeEach(() => {
        processor = new PayProcessor();
        consoleSpy = jest.spyOn(console, "log").mockImplementation();
    });

    afterEach(() => {
        consoleSpy.mockRestore();
    });

    it("should process Credit Card payment", () => {
        processor.process("Credit Card", 100);
        expect(consoleSpy).toHaveBeenCalledWith("Paid 100 using Credit Card");
    });

    it("should process Apple Pay payment", () => {
        processor.process("Apple Pay", 50);
        expect(consoleSpy).toHaveBeenCalledWith("Paid 50 using Apple Pay");
    });

    it("should process Paypal payment", () => {
        processor.process("Paypal", 75);
        expect(consoleSpy).toHaveBeenCalledWith("Paid 75 using Paypal");
    });

    it("should handle invalid payment method", () => {
        processor.process("Bitcoin", 200);
        expect(consoleSpy).toHaveBeenCalledWith("Invalid Payment Method");
    });

    it("should handle empty string as invalid payment method", () => {
        processor.process("", 100);
        expect(consoleSpy).toHaveBeenCalledWith("Invalid Payment Method");
    });
});


// ─── Good Implementation Tests (OCP Compliant) ──────────────────────────────

describe("CreditCardPayment", () => {
    let consoleSpy: jest.SpyInstance;

    beforeEach(() => {
        consoleSpy = jest.spyOn(console, "log").mockImplementation();
    });

    afterEach(() => {
        consoleSpy.mockRestore();
    });

    it("should log correct message when paying with Credit Card", () => {
        const payment = new CreditCardPayment();
        payment.pay(200);
        expect(consoleSpy).toHaveBeenCalledWith("Paid 200 using Credit Card");
    });

    it("should handle zero amount", () => {
        const payment = new CreditCardPayment();
        payment.pay(0);
        expect(consoleSpy).toHaveBeenCalledWith("Paid 0 using Credit Card");
    });
});


describe("PayPalPayment", () => {
    let consoleSpy: jest.SpyInstance;

    beforeEach(() => {
        consoleSpy = jest.spyOn(console, "log").mockImplementation();
    });

    afterEach(() => {
        consoleSpy.mockRestore();
    });

    it("should log correct message when paying with PayPal", () => {
        const payment = new PayPalPayment();
        payment.pay(150);
        expect(consoleSpy).toHaveBeenCalledWith("Paid 150 using PayPal");
    });
});


describe("CryptoPayment", () => {
    let consoleSpy: jest.SpyInstance;

    beforeEach(() => {
        consoleSpy = jest.spyOn(console, "log").mockImplementation();
    });

    afterEach(() => {
        consoleSpy.mockRestore();
    });

    it("should log correct message when paying with Crypto", () => {
        const payment = new CryptoPayment();
        payment.pay(500);
        expect(consoleSpy).toHaveBeenCalledWith("Paid 500 using Crypto");
    });
});


describe("ApplePayPayment (Extension)", () => {
    let consoleSpy: jest.SpyInstance;

    beforeEach(() => {
        consoleSpy = jest.spyOn(console, "log").mockImplementation();
    });

    afterEach(() => {
        consoleSpy.mockRestore();
    });

    it("should log correct message when paying with Apple Pay", () => {
        const payment = new ApplePayPayment();
        payment.pay(300);
        expect(consoleSpy).toHaveBeenCalledWith("Paid 300 using Apple Pay");
    });
});


describe("PaymentProcessor", () => {
    let processor: PaymentProcessor;
    let consoleSpy: jest.SpyInstance;

    beforeEach(() => {
        processor = new PaymentProcessor();
        consoleSpy = jest.spyOn(console, "log").mockImplementation();
    });

    afterEach(() => {
        consoleSpy.mockRestore();
    });

    it("should delegate payment to CreditCardPayment", () => {
        processor.process(new CreditCardPayment(), 200);
        expect(consoleSpy).toHaveBeenCalledWith("Paid 200 using Credit Card");
    });

    it("should delegate payment to PayPalPayment", () => {
        processor.process(new PayPalPayment(), 99);
        expect(consoleSpy).toHaveBeenCalledWith("Paid 99 using PayPal");
    });

    it("should delegate payment to CryptoPayment", () => {
        processor.process(new CryptoPayment(), 1000);
        expect(consoleSpy).toHaveBeenCalledWith("Paid 1000 using Crypto");
    });

    it("should delegate payment to ApplePayPayment", () => {
        processor.process(new ApplePayPayment(), 200);
        expect(consoleSpy).toHaveBeenCalledWith("Paid 200 using Apple Pay");
    });

    it("should work with any custom IPaymentMethod implementation (mock)", () => {
        const mockPayment: IPaymentMethod = { pay: jest.fn() };
        processor.process(mockPayment, 450);
        expect(mockPayment.pay).toHaveBeenCalledWith(450);
    });

    it("should call pay exactly once per process call", () => {
        const mockPayment: IPaymentMethod = { pay: jest.fn() };
        processor.process(mockPayment, 100);
        expect(mockPayment.pay).toHaveBeenCalledTimes(1);
    });
});

File: C#/SOLID.Tests/OCPTest.cs
using System;
using Xunit;
using Moq;

namespace OCP.Tests
{
    // ── 1. CreditCardPayment tests ────────────────────────────────

    [Collection("Sequential")]
    public class CreditCardPaymentTests
    {
        private readonly System.IO.TextWriter _originalOut = Console.Out;

        [Fact]
        public void Pay_WritesCorrectAmountAndMethod()
        {
            var output = new System.IO.StringWriter();
            Console.SetOut(output);

            new CreditCardPayment().Pay(100);

            var result = output.ToString();
            Console.SetOut(_originalOut);   // ← restore BEFORE assert (so failures don't leave Console broken)
            Assert.Contains("100", result);
            Assert.Contains("Credit Card", result);
        }

        [Fact]
        public void CreditCardPayment_ImplementsIPaymentMethod()
        {
            Assert.IsAssignableFrom<IPaymentMethod>(new CreditCardPayment());
        }

        [Theory]
        [InlineData(1)]
        [InlineData(500)]
        [InlineData(9999)]
        public void Pay_VariousAmounts_DoesNotThrow(int amount)
        {
            var output = new System.IO.StringWriter();
            Console.SetOut(output);

            var ex = Record.Exception(() => new CreditCardPayment().Pay(amount));

            Console.SetOut(_originalOut);   // ← always restore
            Assert.Null(ex);
        }
    }
    // ── 2. PayPalPayment tests ────────────────────────────────────

    [Collection("Sequential")]
    public class PayPalPaymentTests
    {
        [Fact]
        public void Pay_WritesCorrectAmountAndMethod()
        {
            var payment = new PayPalPayment();
            using var output = new System.IO.StringWriter();
            Console.SetOut(output);

            payment.Pay(200);

            var result = output.ToString();
            Assert.Contains("200", result);
            Assert.Contains("PayPal", result);
        }

        [Fact]
        public void PayPalPayment_ImplementsIPaymentMethod()
        {
            Assert.IsAssignableFrom<IPaymentMethod>(new PayPalPayment());
        }
    }

    // ── 3. CryptoPayment tests ────────────────────────────────────

    [Collection("Sequential")]
    public class CryptoPaymentTests
    {
        [Fact]
        public void Pay_WritesCorrectAmountAndMethod()
        {
            var payment = new CryptoPayment();
            using var output = new System.IO.StringWriter();
            Console.SetOut(output);

            payment.Pay(300);

            var result = output.ToString();
            Assert.Contains("300", result);
            Assert.Contains("Crypto", result);
        }

        [Fact]
        public void CryptoPayment_ImplementsIPaymentMethod()
        {
            Assert.IsAssignableFrom<IPaymentMethod>(new CryptoPayment());
        }
    }

    // ── 4. ApplePayPayment tests (the OCP extension) ──────────────

    [Collection("Sequential")]
    public class ApplePayPaymentTests
    {
        [Fact]
        public void Pay_WritesCorrectAmountAndMethod()
        {
            var payment = new ApplePayPayment();
            using var output = new System.IO.StringWriter();
            Console.SetOut(output);

            payment.Pay(400);

            var result = output.ToString();
            Assert.Contains("400", result);
            Assert.Contains("Apple Pay", result);
        }

        [Fact]
        public void ApplePayPayment_ImplementsIPaymentMethod()
        {
            // OCP: new payment type added by extension, not modifying existing code
            Assert.IsAssignableFrom<IPaymentMethod>(new ApplePayPayment());
        }
    }

    // ── 5. PaymentProcessor tests ─────────────────────────────────

    public class PaymentProcessorTests
    {
        // Uses Moq — no Console involved, safe to run in parallel

        [Fact]
        public void Process_CallsPayOnce_WithCorrectAmount()
        {
            var mockPayment = new Mock<IPaymentMethod>();
            var processor = new PaymentProcessor();

            processor.Process(mockPayment.Object, 150);

            mockPayment.Verify(p => p.Pay(150), Times.Once);
        }

        [Fact]
        public void Process_CalledMultipleTimes_CallsPayEachTime()
        {
            var mockPayment = new Mock<IPaymentMethod>();
            var processor = new PaymentProcessor();

            processor.Process(mockPayment.Object, 100);
            processor.Process(mockPayment.Object, 200);
            processor.Process(mockPayment.Object, 300);

            mockPayment.Verify(p => p.Pay(It.IsAny<int>()), Times.Exactly(3));
        }

        [Theory]
        [InlineData(0)]
        [InlineData(50)]
        [InlineData(1000)]
        public void Process_VariousAmounts_ForwardsExactAmountToPaymentMethod(int amount)
        {
            var mockPayment = new Mock<IPaymentMethod>();
            var processor = new PaymentProcessor();

            processor.Process(mockPayment.Object, amount);

            mockPayment.Verify(p => p.Pay(amount), Times.Once);
        }
    }

    // ── 6. OCP Compliance tests — the principle itself ────────────

    public class OCPComplianceTests
    {
        [Fact]
        public void AllPaymentTypes_ImplementIPaymentMethod()
        {
            // Every payment class satisfies the same interface —
            // PaymentProcessor never needs to change when new ones are added
            Assert.IsAssignableFrom<IPaymentMethod>(new CreditCardPayment());
            Assert.IsAssignableFrom<IPaymentMethod>(new PayPalPayment());
            Assert.IsAssignableFrom<IPaymentMethod>(new CryptoPayment());
            Assert.IsAssignableFrom<IPaymentMethod>(new ApplePayPayment());
        }

        [Fact]
        public void PaymentProcessor_WorksWithAnyIPaymentMethod_WithoutModification()
        {
            // OCP core: processor accepts any payment type through the abstraction
            // Adding ApplePayPayment required zero changes to PaymentProcessor
            var processor = new PaymentProcessor();
            IPaymentMethod[] allPayments =
            {
                new CreditCardPayment(),
                new PayPalPayment(),
                new CryptoPayment(),
                new ApplePayPayment()
            };

            foreach (var payment in allPayments)
            {
                var ex = Record.Exception(() => processor.Process(payment, 100));
                Assert.Null(ex);
            }
        }

        [Fact]
        public void IPaymentMethod_HasExactlyOneMethod_Pay()
        {
            // The interface contract is minimal and stable —
            // new payment types extend behaviour without touching this
            var methods = typeof(IPaymentMethod).GetMethods();
            Assert.Single(methods);
            Assert.Equal("Pay", methods[0].Name);
        }
    }
}